agent/local/state.go

```diff
@@ -1251,9 +1346,14 @@ func (l *State) syncCheck(key structs.CheckID) error {
 
 	var out struct{}
 	err := l.Delegate.RPC("Catalog.Register", &req, &out)
+	l.Lock()
+	defer l.Unlock()
 	switch {
 	case err == nil:
 		l.checks[key].InSync = true
+		if l.checks[key].version > currentVersion {
+			l.checks[key].InSync = false
+		}
 		// Given how the register API works, this info is also updated
 		// every time we sync a check.
 		l.nodeInfoInSync = true
```

Comment:
There is a case with `acl` stuff below, which also needs to check the key before setting `InSync`, because it could gone.